File: android/app/src/main/java/ai/strainease/app/ui/checkin/CheckInForm.kt
package ai.strainease.app.ui.checkin

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.OutlinedTextFieldDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import ai.strainease.app.data.CheckIn
import ai.strainease.app.data.CheckInMetrics
import ai.strainease.app.data.CheckInStore
import kotlinx.coroutines.launch

/**
 * Form for today's check-in. Four 1-5 scales, optional note,
 * and a "Clear" affordance so the patient can wipe the day if
 * they logged by mistake. Direct port of the iOS
 * `CheckInForm.swift` shape and the web `CheckInForm.tsx`.
 *
 * Reads the current entry from the provided [today] prop so
 * the form can hydrate without a second DataStore read; the
 * parent is responsible for keeping it in sync with the
 * [CheckInStore] flow.
 *
 * The form owns the local `metrics` and `note` state so the
 * save callback can be wired directly to the [store] without
 * round-tripping through the parent.
 */
@Composable
fun CheckInForm(
    today: CheckIn?,
    store: CheckInStore,
    modifier: Modifier = Modifier,
) {
    var metrics by remember { mutableStateOf(CheckInStore.defaultMetrics) }
    var note by remember { mutableStateOf("") }
    var isSaving by remember { mutableStateOf(false) }
    val scope = rememberCoroutineScope()

    LaunchedEffect(today?.date, today?.updatedAt) {
        if (today != null) {
            metrics = today.metrics
            note = today.note
        } else {
            metrics = CheckInStore.defaultMetrics
            note = ""
        }
    }

    val scales = listOf(
        ScaleSpec(label = "Mood", hint = "1 = awful, 5 = great", highIsGood = true),
        ScaleSpec(label = "Sleep", hint = "1 = none, 5 = fully rested", highIsGood = true),
        ScaleSpec(label = "Pain", hint = "1 = none, 5 = severe", highIsGood = false),
        ScaleSpec(label = "Anxiety", hint = "1 = calm, 5 = severe", highIsGood = false),
    )

    Column(verticalArrangement = Arrangement.spacedBy(16.dp), modifier = modifier) {
        scales.forEach { scale ->
            scaleRow(scale, metrics, onPick = { value ->
                metrics = when (scale.label) {
                    "Mood" -> metrics.copy(mood = value)
                    "Sleep" -> metrics.copy(sleep = value)
                    "Pain" -> metrics.copy(pain = value)
                    "Anxiety" -> metrics.copy(anxiety = value)
                    else -> metrics
                }
            })
        }

        Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
            Text(
                "Note (optional)",
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            OutlinedTextField(
                value = note,
                onValueChange = { v -> if (v.length <= CheckInStore.CHECK_IN_NOTE_MAX) note = v },
                placeholder = { Text("Anything to remember?") },
                modifier = Modifier.fillMaxWidth(),
                minLines = 3,
                colors = fieldColors(),
            )
            Text(
                "${note.length} / ${CheckInStore.CHECK_IN_NOTE_MAX}",
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }

        Row(
            horizontalArrangement = Arrangement.spacedBy(10.dp),
            modifier = Modifier.fillMaxWidth(),
        ) {
            SaveButton(
                isLoading = isSaving,
                onClick = {
                    if (isSaving) return@SaveButton
                    isSaving = true
                    scope.launch {
                        store.upsert(metrics, note)
                        isSaving = false
                    }
                },
                modifier = Modifier.weight(1f),
            )
            if (today != null) {
                ClearButton(
                    onClick = {
                        if (isSaving) return@ClearButton
                        isSaving = true
                        scope.launch {
                            store.delete(CheckInStore.todayKey())
                            metrics = CheckInStore.defaultMetrics
                            note = ""
                            isSaving = false
                        }
                    },
                )
            }
        }
    }
}

private data class ScaleSpec(
    val label: String,
    val hint: String,
    val highIsGood: Boolean,
)

@Composable
private fun scaleRow(
    scale: ScaleSpec,
    metrics: CheckInMetrics,
    onPick: (Int) -> Unit,
) {
    val current = when (scale.label) {
        "Mood" -> metrics.mood
        "Sleep" -> metrics.sleep
        "Pain" -> metrics.pain
        "Anxiety" -> metrics.anxiety
        else -> 3
    }
    Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier.fillMaxWidth(),
        ) {
            Text(
                scale.label,
                style = MaterialTheme.typography.bodyMedium,
                fontWeight = FontWeight.SemiBold,
            )
            Box(Modifier.weight(1f))
            Text(
                scale.hint,
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
            (1..5).forEach { i ->
                val selected = current == i
                Box(
                    modifier = Modifier
                        .size(38.dp)
                        .clip(CircleShape)
                        .background(
                            if (selected) scaleColor(i, scale.highIsGood) else MaterialTheme.colorScheme.surface
                        )
                        .border(1.dp, MaterialTheme.colorScheme.outline, CircleShape)
                        .clickable { onPick(i) },
                    contentAlignment = Alignment.Center,
                ) {
                    Text(
                        "$i",
                        color = if (selected) Color.White else MaterialTheme.colorScheme.onSurface,
                        fontWeight = FontWeight.SemiBold,
                    )
                }
            }
        }
    }
}

private fun scaleColor(value: Int, highIsGood: Boolean): Color {
    val goodEnd = if (highIsGood) 5 else 1
    val distance = kotlin.math.abs(value - goodEnd)
    return when {
        distance <= 1 -> Color(0xFF22C55E) // emerald-500
        distance <= 2 -> Color(0xFFF59E0B) // amber-500
        else -> Color(0xFFEF4444) // rose-500
    }
}

@Composable
private fun fieldColors() = OutlinedTextFieldDefaults.colors(
    focusedContainerColor = MaterialTheme.colorScheme.surface,
    unfocusedContainerColor = MaterialTheme.colorScheme.surface,
    focusedBorderColor = MaterialTheme.colorScheme.outline,
    unfocusedBorderColor = MaterialTheme.colorScheme.outline,
    focusedTextColor = MaterialTheme.colorScheme.onSurface,
    unfocusedTextColor = MaterialTheme.colorScheme.onSurface,
    cursorColor = MaterialTheme.colorScheme.primary,
)

@Composable
private fun SaveButton(
    isLoading: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier
            .clip(RoundedCornerShape(50))
            .background(MaterialTheme.colorScheme.primary)
            .clickable(enabled = !isLoading, onClick = onClick)
            .padding(horizontal = 16.dp, vertical = 10.dp),
        horizontalArrangement = Arrangement.Center,
        verticalAlignment = Alignment.CenterVertically,
    ) {
        if (isLoading) {
            CircularProgressIndicator(
                modifier = Modifier.size(16.dp),
                strokeWidth = 2.dp,
                color = MaterialTheme.colorScheme.onPrimary,
            )
        } else {
            Icon(
                imageVector = Icons.Default.Check,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.onPrimary,
                modifier = Modifier.size(16.dp),
            )
        }
        Text(
            "  Save",
            color = MaterialTheme.colorScheme.onPrimary,
            fontWeight = FontWeight.SemiBold,
            fontSize = 14.sp,
        )
    }
}

@Composable
private fun ClearButton(
    onClick: () -> Unit,
) {
    Row(
        modifier = Modifier
            .clip(RoundedCornerShape(50))
            .background(MaterialTheme.colorScheme.surface)
            .border(1.dp, MaterialTheme.colorScheme.outline, RoundedCornerShape(50))
            .clickable(onClick = onClick)
            .padding(horizontal = 16.dp, vertical = 10.dp),
        horizontalArrangement = Arrangement.Center,
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Icon(
            imageVector = Icons.Default.Delete,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.onSurface,
            modifier = Modifier.size(16.dp),
        )
        Text(
            "  Clear",
            color = MaterialTheme.colorScheme.onSurface,
            fontWeight = FontWeight.Medium,
            fontSize = 14.sp,
        )
    }
}

File: ios/StrainEase/Account/CheckInForm.swift
// Form for today's check-in. Four 1-5 scales, optional note,
// and a "Clear" affordance so the patient can wipe the day if
// they logged by mistake. Direct port of the web
// `CheckInForm.tsx` shape.

import SwiftUI

struct CheckInForm: View {
    @Environment(CheckInStore.self) private var store

    @State private var metrics: CheckInMetrics = defaultCheckInMetrics
    @State private var note: String = ""
    @State private var didHydrate = false
    @State private var isSaving = false

    /// `today` is the patient's current check-in for today, or
    /// `nil` if they haven't logged one yet. `onSaved` lets the
    /// panel collapse after a successful write.
    var today: CheckIn?
    var onSaved: (() -> Void)?

    private let scales: [(key: WritableKeyPath<CheckInMetrics, Int>, label: String, hint: String, highIsGood: Bool)] = [
        (\.mood, "Mood", "1 = awful, 5 = great", true),
        (\.sleep, "Sleep", "1 = none, 5 = fully rested", true),
        (\.pain, "Pain", "1 = none, 5 = severe", false),
        (\.anxiety, "Anxiety", "1 = calm, 5 = severe", false),
    ]

    private var noteBinding: Binding<String> {
        Binding(
            get: { String(note.prefix(CheckInStore.checkInNoteMax)) },
            set: { note = String($0.prefix(CheckInStore.checkInNoteMax)) }
        )
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            ForEach(scales, id: \.label) { scale in
                scaleRow(scale)
            }
            VStack(alignment: .leading, spacing: 6) {
                Text("Note (optional)")
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundStyle(Palette.mutedForeground)
                TextField(
                    "Anything to remember?",
                    text: noteBinding,
                    axis: .vertical
                )
                .lineLimit(3, reservesSpace: true)
                .padding(10)
                .background(Palette.card, in: RoundedRectangle(cornerRadius: 10))
                .overlay(
                    RoundedRectangle(cornerRadius: 10)
                        .strokeBorder(Palette.border, lineWidth: 1)
                )
                .font(.system(size: 14))
                Text("\(note.count) / \(CheckInStore.checkInNoteMax)")
                    .font(.system(size: 11))
                    .foregroundStyle(Palette.mutedForeground)
            }
            HStack(spacing: 10) {
                Button {
                    Task { await save() }
                } label: {
                    HStack(spacing: 6) {
                        if isSaving {
                            ProgressView()
                                .controlSize(.small)
                        } else {
                            Image(systemName: "checkmark")
                        }
                        Text("Save")
                    }
                    .font(.system(size: 14, weight: .semibold))
                    .padding(.horizontal, 16)
                    .padding(.vertical, 10)
                    .frame(maxWidth: .infinity)
                    .foregroundStyle(Palette.primaryForeground)
                    .background(Palette.primary, in: Capsule())
                }
                .buttonStyle(.plain)
                .disabled(isSaving)

                if today != nil {
                    Button(role: .destructive) {
                        Task { await clear() }
                    } label: {
                        HStack(spacing: 6) {
                            Image(systemName: "trash")
                            Text("Clear")
                        }
                        .font(.system(size: 14, weight: .medium))
                        .padding(.horizontal, 16)
                        .padding(.vertical, 10)
                        .foregroundStyle(Palette.foreground)
                        .background(Palette.card, in: Capsule())
                        .overlay(
                            Capsule().strokeBorder(Palette.border, lineWidth: 1)
                        )
                    }
                    .buttonStyle(.plain)
                    .disabled(isSaving)
                }
            }
        }
        .onAppear { hydrateIfNeeded() }
        .onChange(of: today?.updatedAt) { _, _ in hydrateIfNeeded() }
    }

    private func scaleRow(_ scale: (key: WritableKeyPath<CheckInMetrics, Int>, label: String, hint: String, highIsGood: Bool)) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            HStack {
                Text(scale.label)
                    .font(.system(size: 14, weight: .semibold))
                Spacer()
                Text(scale.hint)
                    .font(.system(size: 11))
                    .foregroundStyle(Palette.mutedForeground)
            }
            HStack(spacing: 6) {
                ForEach(1...5, id: \.self) { i in
                    let selected = metrics[keyPath: scale.key] == i
                    Button {
                        metrics[keyPath: scale.key] = i
                    } label: {
                        Text("\(i)")
                            .font(.system(size: 14, weight: .semibold))
                            .frame(width: 38, height: 38)
                            .foregroundStyle(selected ? Palette.primaryForeground : Palette.foreground)
                            .background(
                                selected ? colorForMetric(i, highIsGood: scale.highIsGood) : Palette.card,
                                in: Circle()
                            )
                            .overlay(
                                Circle().strokeBorder(
                                    selected ? Color.clear : Palette.border,
                                    lineWidth: 1
                                )
                            )
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel("\(scale.label) \(i) of 5")
                }
            }
        }
    }

    private func colorForMetric(_ value: Int, highIsGood: Bool) -> Color {
        let goodEnd = highIsGood ? 5 : 1
        let distance = abs(value - goodEnd)
        if distance <= 1 { return .green }
        if distance <= 2 { return .orange }
        return .red
    }

    private func hydrateIfNeeded() {
        guard !didHydrate || today?.updatedAt != nil else {
            if let today {
                metrics = today.metrics
                note = today.note
            }
            didHydrate = true
            return
        }
        if let today {
            metrics = today.metrics
            note = today.note
        } else {
            metrics = defaultCheckInMetrics
            note = ""
        }
        didHydrate = true
    }

    private func save() async {
        isSaving = true
        defer { isSaving = false }
        await store.upsertToday(metrics: metrics, note: note)
        if store.errorMessage == nil {
            onSaved?()
        }
    }

    private func clear() async {
        isSaving = true
        defer { isSaving = false }
        await store.delete(dateId: CheckInStore.todayKey())
    }
}

#Preview("Empty") {
    SWCard {
        VStack(alignment: .leading, spacing: 12) {
            SectionLabel("Daily check-in")
            CheckInForm(today: nil)
        }
    }
    .padding()
    .environment(CheckInStore.preview([]))
}

#Preview("With Today") {
    let today = CheckIn(
        id: CheckInStore.todayKey(),
        date: CheckInStore.todayKey(),
        metrics: CheckInMetrics(mood: 4, sleep: 3, pain: 2, anxiety: 2),
        note: "Felt good after a long walk.",
        createdAt: 1_700_000_000_000,
        updatedAt: 1_700_000_000_000
    )
    return SWCard {
        VStack(alignment: .leading, spacing: 12) {
            SectionLabel("Daily check-in")
            CheckInForm(today: today)
        }
    }
    .padding()
    .environment(CheckInStore.preview([today]))
}
